### ISSUE TITLE
[Question] em.isActive(entityId)

### ISSUE THREAD

**Issue by DaanVanYperen** [junkdog/artemis-odb#573]
@junkdog implementing the testcases for the debugger I'm running into some engine behavior I can't find any documentation on.

After entity `deleteFromWorld` is issued, until deletion is finalized by the `batchprocessor`, `em.isActive(entityId)` returns `true`. Does this mean all interactions with the entity (except for deletion) are considered legal? Or should the method return `false` right after `deleteFromWorld`?

I ask since it seems like an artifact from how the engine works and not so much intended (see `EntityTransmuter#isValid` for example, it will block transmutations which will block componentmappers from triggering create/delete). But for some reason it treats deleted-finalized entities differently and throw an exception.

Edit: `DelayedComponentRemoval` suggests component read is intended for that group until all subscriptions have been notified.

**Comment by junkdog** [comment]
> After entity `deleteFromWorld` is issued, until deletion is finalized by the `batchprocessor`, `em.isActive(entityId)` returns `true`. Does this mean all interactions with the entity (except for deletion) are considered legal? Or should the method return `false` right after `deleteFromWorld`?

It's a bit ugly. We basically have 2 deleted timelines: one works on the EntitySubscription layer, inserting/removing (primarily) to/from systems. `em.isActive` until purged by EntityManaged.

World.process() is considered atomic in the sense that no entity id:s are reused during the same process round - this avoids headaches + allows acting on deleted entities in later systems (limited, but if an entityId is all which is required or `DelayedComponentRemoval` is all you need, it's there).

Ideally, EntitySubscriptions should only fire before and after all systems have processed. It makes more sense to think of the entity state as a snapshot view for the active process - the system-level granularity of EntitySubscriptions introduces bugs frin complicating glow + accidental entity creation "too late" into the system loop.

If the two were one; there'd just be one flow to reason about.

The tricky part is that this impacts removal of components. They essentially have to be delayed too - if EntitySubscriptions tick with the process and not systems. For consistency, creation of components should maybe also be delayed.

A tricky edge-case:
 1) system1  removes ComponentA
 2) system2: creates ComponentA

If remove is delayed, it shouldn't be undone by system1:s componentA.remove.

> I ask since it seems like an artifact from how the engine works and not so much intended (see `EntityTransmuter#isValid` for example, it will block transmutations which will block componentmappers from triggering create/delete).

I think that's a silly optimization; I should probably have a talk with my former self.


> Edit: `DelayedComponentRemoval` suggests component read is intended for that group until all subscriptions have been notified.

Yes, when world.delete(e) is issued - all component removal happen when the entity is purged at the end of process(). Mapper.remove removes immediately unless they come with `@DelayedComponentRemoval`

**Comment by DaanVanYperen** [comment]
This might turn into a bit of a religious discussion, so sorry about that beforehand ;) You're a lot more read up on ECS. 

I always thought of Systems as atomic operations in a pipeline. As a developer I have to be able to trust the data that goes into the system is current. If I can't, if other systems might have performed operations that I won't be aware of it would mean I have to consider implementation details of other systems, which seems undermines one of the big benefits of ECS.

That the engine puts certain restrictions on developers is fine:
- Subscriptions frozen during system.process: easy to reason about.
- Certain updates delayed until right the `system.process` finishes; fine.

All devs need is for it to be consistent and well documented, perhaps be a bit more intuitive as well. Making `DelayedComponentRemoval=true` the default would help.

The main reason I asked is because I have to code some rules into the debugger. Right now I don't know when `ComponentMapper` interactions are *intended* to be legal. (Can divine it's coincides with DelayedComponentRemoval is true).

**Comment by junkdog** [comment]
> I always thought of Systems as atomic operations in a pipeline.

I think we're saying the same thing. Systems are atomic too, what I mean is that entity subscriptions are only updated before and after systems.process. It introduces a potential delay of 1 frame, but for all practical purposes - it should still happen in the same frame (say, act on input before processing frame).

The entity-system contract remains, but component deletion is scheduled until the end of process(). No changes should be necessary to systems - except if a previous system depended on inserting entities during the same fram (for this to work in practice, one has know when it's safe to do so - and remember that when updating the systems list)

> Making `DelayedComponentRemoval=true` the default would help.

This is what I intend, but going all the way and disallowing the current process. It's quite the refactoring though, but would simplify a lot of the internals + present everything in a more consistent manner.

**Comment by DaanVanYperen** [comment]
So easy to confuse things with everything being called process. 

`The entity-system contract remains, but component deletion is scheduled until the end of process().`

Before:
component deletion: immediatly, or post `System#process`
System#remove: post `System#process`
entity deletion: post `System#process`
subscriptions updated: post `Systems#process`

After:
component deletion: mid or post `System#process`
System#remove: post `System#process`
entity deletion: post `System#process`
subscriptions updated: post `**World**#process`

?

**Comment by junkdog** [comment]
Kinda:

what|Current Impl | Future
----|-------------- | -------------
component deletion | between systems or `@DelayedComponentRemoval`| Similar to `@DelayedComponentRemoval`
subscriptions updated | between systems, and on world.process() | first and last in world.process()
entity deletion | between systems, purge last on world.process() end |  first and last in world.process()
subscriptions updated | between systems, and on world.process() |  first and last in world.process()

Systems::remove is a SubscriptionListener, so same as for subscriptions.

**Comment by DaanVanYperen** [comment]
I agree with most of that except for the subscriptions. The future way of handling subscriptions would essentially extend the side effects of a system all the way till the end of `world.process`, instead of just keep them system local.

In the current impl, I can trust subscriptions to give me the entities of the specified composition, unless i break something locally in the system. In the future version, I'll have to take implementation details of previous systems into account. (=coding really defensively).

The flow of operations becomes harder to follow as well. Some systems will trigger immediately because the composition happens to match, while others get pushed one or more frames ahead. So what looks like A->B->C becomes A->C->B.

**Comment by junkdog** [comment]
> I agree with most of that except for the subscriptions. The future way of handling subscriptions would essentially extend the side effects of a system all the way till the end of `world#process`, instead of just keep them system local.

But the subscriptions remains valid for the duration of `World::process`. It's just that component and entity deletion is scheduled for the end.



> The flow of operations becomes harder to follow as well. Some systems will trigger immediately because the composition happens to match,

Ah, no - EntitySubscription:s listeners fire in tandem with the state changes.

Mapper.create() is thus only visible at the callsite where it was issued, and for the rest - when entity subscriptions are updated.

**Comment by DaanVanYperen** [comment]
God I hate discussing this via text. :p 

So if I kill Yak A in system 1, spawn a Yak B in system 2, and shave Yaks in system 3,  System 3 will shave the dead yak and ignore the new yak?

**Comment by junkdog** [comment]
> So if I kill Yak A in system 1, spawn a Yak B in system 2, and shave Yaks in system 3, System 3 will shave the dead yak and ignore the new yak?

It's a fancy passing. The new yak is visible to System 3 the subsequent `World::process` round.

> Edit: Say I have a `BitmapAsset`, and I have a `BitmapAssetManager::remove` that releases `BitMapAsset` texture. my BitmapAssetRenderer would still be able to access `BitmapAsset`?

Hmm, as long as `BitmapAssetManager::remove` is an EntitySubscription listener and `BitmapAssetRenderer::processSystem` renders the entity - then yes. If so, yes - BitmapAsset would be accessible until the EntitySubscription listeners have fired (the next step would be to remove the components, delete any entities).

And then on the next round, the entity is evicted.

....

Or wait, you mean you remove the component with BitmapAssetManager::remove, during the same world::process - before rendering? If so - remove component in BitmapAssetManager::remove, and do the acquire/release in an EntitySubscription (so as to sync the asset lifetime with World::process).

**Comment by DaanVanYperen** [comment]
We're still not synced up I think. Bit at a loss.

**Comment by DaanVanYperen** [comment]
>> I agree with most of that except for the subscriptions. The future way of handling subscriptions would essentially extend the side effects of a system all the way till the end of world#process, instead of just keep them system local.

> But the subscriptions remains valid for the duration of World::process. It's just that component and entity deletion is scheduled for the end.

It's like we have a database on which we operate in series and now want to make operations effectively run in parallel. That's never going to happen without putting a lot of burden on the operation creators and having some amazing merge technology at the end.

**Comment by junkdog** [comment]
> It's like we have a database on which we operate in series and now want to make operations effectively run in parallel. That's never going to happen without putting a lot of burden on the operation creators and having some amazing consolidation technology at the end.

Hmm, component data still mutate sequentially - but compositionId remains the same (and is valid) for each entity until end of World::process. One problem with the current implementation is that entities created by systems are never processed by preceding systems during their first round, so one has to be mindful about how/when entities are created. This applies doubly to when systems are re-ordered.

I'm not sure I follow how this creates extra complexity? Part of the thinking is to remove complexity - have everything beat to the same rhythm, but maybe I'm missing a use-case?

From a high-level, conceptually it becomes:
World process
  -> create entities
  -> create components
  -> process systems
  -> delete components
  -> mutate entities
  -> delete entities

While not 100% accurate, it's true enough from the viewpoint of the user: creation happens before systems, deletion and mutation after systems. 

> Edit:

Good catch. Yeah, this wouldn't work. One would either have to resolve that in PetaAdoptSystem::inserted (during the EntitySubscription update phase, componentId:s are allowed to update repeatedly) or introduce a secondary component/`Adopted::state`.

**Comment by junkdog** [comment]
(going out / back tomorrow)

**Comment by DaanVanYperen** [comment]
Just trying to challenge the idea, not looking for a fight. ;)

> the current implementation is that entities created by systems are never processed by preceding systems during their first round, so one has to be mindful about how/when entities are created. This applies doubly to when systems are re-ordered.

Aspects make it so I can guarantee the system only gets what it can handle, or you implement an `EntitySpawnerSystem`. I can imagine it differs from user to user.

> I'm not sure I follow how this creates extra complexity? 

I'd personally struggle with it. It's less intuitive than it looks imo. it feels a bit like the pivot from OOP to ECS, not sure if users will manage to grok it. 

**Stale data & Multi-frame initialization**

Stale data is my biggest gripe. Combine that with multi-frame initialization this'll make it hard to discern flow and reason about and can end up running multiple frames.

Take this simple example:

  1. `PeonSpawnSystem` -> spawn peon. 
  2. `RandomJobSystem @All(Peon.class)`  -> randomly become farmer or warrior each frame.
  3. `SoldiersAiSystem @All(Sword.class,Peon.class)` -> attack enemy with held.
  4. `FarmerAiSystem @All(Hoe.class,Peon.class)` -> plant what is held.
  5. `PeonRenderSystem @All(Peon.class)`

**Old system:**
   Peon gets spawned, is a farmer, plants some grain. 
 16 ms 

In the old system users can just look at the systems and the aspects, and reason what will happen.

**New system**
  on frame 1, Peon gets spawned. Nothing else happens.
  on frame 2, Peon becomes a farmer.  Not rendered as one though.
  on frame 3, Peon turns into a warrior again, but runs farmer AI and he plants a sword into a field. Rendered as a farmer.
  on frame 4, Peon turns into a farmer, but runs warrior AI and he stabs someone with grain. 
  66ms

In the new system, order of systems means nothing composition wise, only data wise, aspects mean nothing because systems get stale data, and while coding the user constantly have to make the mental gymnastics 'but how will this be smeared out over multiple frames. This aspect says A but i'm really getting B'. Also get some nice lag for free.

Imagine that with a 80 system game. Or with systems that do not operate each frame. Arguably you could probably design your game around the future pattern, but in the old system, it all works very intuitively.

This is a very broad problem IMO, to replace a very narrow one (moment of entity creation).

There are more concerns (behavior change in listeners, order of operations on components from system to system).

I wouldn't mind a toggle for this functionality. Basically could be two different invocation strategies.

**Comment by DaanVanYperen** [comment]
I asked https://talktotransformer.com/ 

![image](https://user-images.githubusercontent.com/6305602/57951934-ecfa8700-78eb-11e9-84e2-809a1224942f.png)

**Comment by junkdog** [comment]
> Just trying to challenge the idea, not looking for a fight. ;)

No, this is interesting - I want us to be one the same page, there are obviously use-cases I haven't considered + views that have a different perspective. If the idea is flawed, it should not be implemented.

I'll expand tomorrow + include use-cases I'm trying to solve, when I return home.

**Comment by DaanVanYperen** [comment]
> No, this is interesting - I want us to be one the same page, there are obviously use-cases I haven't considered + views that have a different perspective. If the idea is flawed, it should not be implemented.

It's your project so I'm happy to follow. It's a religious discussion in the same way most IntellIJ vs Eclipse discussions end up. I'm sure there must be ECS out there that do it this way, or a variant of it.

Tweaking the engine to be more flexible in how it handles lifecycle is not a bad thing, providing `DelayedEntityCreationInvocationStrategy` initially would allow people to PoC with it. Think for games where every frame counts (FPS/bullet hell etc) it makes sense keeping the old way around as well.

The debugger can benefit from this as well. I would like to extend the 'graveyard' for entity IDs beyond a few frames, so it becomes easier to detect dangling references.

**Comment by nanonull** [comment]
Guys, can I add my small opinion about 'entity processing lasts several frames when I'd want it took 1 frame'.

Sometimes logic in my game requires something like sub-world. I have a lot systems which calculates (invalidates/validates) unit parameters depending on unit position, attributes, effects, inventory, effects of inventory items, adjacent units and much more. Even worse, sometimes unit calculation  may invalidate another unit, so another entity will enter into hell logic loop. 

So unit calculation may take from 1 to x frames depending on how many changes (invalidations) unit has.

Then I moved all this unit logic into another world (unit-world) with own systems. Unit-world completes unit calculation in current frame. 
It seams it works, the only problem is when one unit invalidates another (which said he is valid before) - this means he has to repeat validation on next frame of main-world... And maybe he will invalidate another unit...

I mean things may go hell crap sometimes.

**Comment by DaanVanYperen** [comment]
For the debugger it is now implemented as:
- `@DelayedComponentRemoval ` components are legal to interact with, until engine deletes entity.
- The only exception is adding components to the entity post deletion, that is treated as bad practice / error.

**Comment by DaanVanYperen** [comment]
Question answered. Feel free to create a new ticket for any insights gained during this discussion! Closing.
